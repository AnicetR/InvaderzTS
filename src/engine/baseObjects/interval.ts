import { intervalInterface } from "../types/intervalTypes";

export class interval implements intervalInterface{
    _id: number;

    constructor(readonly each: number,
                readonly callback: CallableFunction){}

    start(){
        if(!this._id){
            this._id = setInterval(
                this.callback,
                this.each
            );
        }else{
            throw new Error('Timeout already running');
        }
    }

    stop(){
        if(this._id){
            clearTimeout(this._id)
            this._id = null;
        }
    }
}