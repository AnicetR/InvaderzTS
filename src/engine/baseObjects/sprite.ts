import { position } from "../types/gameObjectTypes";

/**
 * Sprite object
 */

export class Sprite{
    
    /**
     * Sprite position on the canvas
     */
    public position: position = {
        x: 0,
        y: 0
    };

    /**
     * Is the sprite image loaded in browser ?
     */
    public loaded: boolean = false;

    public image: HTMLImageElement;

    /**
     * 
     * @param imagePath local path of the image to show
     */
    constructor(private readonly imagePath: string){
        this.image = new Image;
        this.image.src = imagePath;
        this.image.onload = () => {
            this.loaded = true;
        }
    }

}