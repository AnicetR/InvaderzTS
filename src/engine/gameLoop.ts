import { Renderer } from "./renderer/renderer";
import { fpsHistoryStore } from "./stores/debug/fpsHistoryStore";

export class gameLoop{

    

    //The game loop will be iterated 30 times per second
    private _loopIterationsPerSecond: number = 30;
    private _timeStep: number = 1000/this._loopIterationsPerSecond;
    
    private _isPaused: boolean = false;
    private _delta: number = 0;

    //debug store and stuff
    private _frameStore: fpsHistoryStore;


    /**
     * requestAnimationFrame window callback with polyfill
     */
    private readonly _requestAnimationFrame: CallableFunction = 
        (window.requestAnimationFrame        ||
        window.webkitRequestAnimationFrame  ||
        function(callback: CallableFunction){
            window.setTimeout( callback(Date.now()), 1000/this._loopIterationsPerSecond );
        }).bind(window);
    

    constructor(private readonly _renderer: Renderer){
        this._frameStore = fpsHistoryStore.getInstance();
    }

    start() : void{
        this.loop(performance.now());
    }

    private loop(currentTick: number) : void
    {   
        //Don't do anything if the render is ahead of his time
        if(currentTick < (this._renderer.lastTick + this._timeStep)){
            this._requestAnimationFrame(this.loop.bind(this));
            return;
        }

        //Delta definition in case the tick is in advance on frames rendering
        this._delta = currentTick - this._renderer.lastTick;

        //Let's update with the delta, to catch up on render time
        while(this._delta >= this._timeStep){
            this.updateLayers();
            this._delta -= this._timeStep;

            //This is like a realy simple not realy right method to do some interpolation
            //So the gameObject Properties are updated a last time with the left delta as factor
            if(this._delta < this._timeStep){
                this.updateLayers();
            }
        }
        
        //Render
        this.launchRender(currentTick);
    }

    private launchRender(currentTick: number): void
    {
        this._frameStore.fpsLog(currentTick, this._renderer.lastTick);
        this._renderer.render(currentTick);
        this._requestAnimationFrame(this.loop.bind(this));
    }

    private updateLayers(): void
    {
        this._renderer.layers.forEach((layer) => layer.update(this._delta));
    }

}