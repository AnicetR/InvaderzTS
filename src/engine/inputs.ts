/**
 * Inputs Manages keyboard inputs.
 */
export class Inputs {

    /**
     * Current keys down
     */
    public readonly keyCodesDown: Array<number> = [];

    /**
     * defines document based callbacks
     */
    constructor(){
        document.onkeydown = (event) => {
            const keyCode: number = event.keyCode;
            const keyCodeIndex: number = this.keyCodesDown.indexOf(keyCode);
            if(keyCodeIndex == -1){
                this.keyCodesDown.push(keyCode)
            }
        }

        document.onkeyup = (event) => {
            const keyCode: number = event.keyCode;
            const keyCodeIndex: number = this.keyCodesDown.indexOf(keyCode);
            if(keyCodeIndex >= -1){
                this.keyCodesDown.splice(keyCodeIndex, 1);
            }
        }
    }

    /**
     * Check if a specific key is down or not
     * @param keyCode 
     */
    isKeyDown(keyCode: number) : boolean {
        return this.keyCodesDown.includes(keyCode);
    }

}