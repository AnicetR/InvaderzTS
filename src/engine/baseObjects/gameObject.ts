
import { position, gameObjectInterface } from "../types/gameObjectTypes";
import { Sprite } from "./sprite";
import { Context } from "../context";
import { RenderLayer } from "../renderer/renderLayer";
import { collisionManager } from "../collisions/collisionManager";
import { collisionBoxInterface } from "../types/collisionTypes";
import { rectCollisionBox } from "../collisions/rectCollisionBox";

/**
 * gameObject base class
 */
export class gameObject implements gameObjectInterface{

    /**
     * uuid of the object
     */
    uuid: string;   
    
    /**
     * RenderLayer in wich the gameObject is rendered
     */
    renderLayer: RenderLayer;
    
    /**
     * x-y position of the object
     */
    position: position = {
        x: 0,
        y: 0
    };

    /**
     * The sprite of the object
     */
    sprite: Sprite;

    /**
     * The sprite Collection of the object
     */
    spriteCollection: Array<Sprite>;

    /**
     * is the object registered in a layer ?
     */
    private registered: boolean = false;

    collisionBox: collisionBoxInterface;

    constructor(){
        if(this.collisionBox){
            collisionManager.instance.addCollision(this.collisionBox);
        }
    }

    /**
     * Callback called before draw method in the rendering process
     */
    update(delta: number): void {
        if(this.sprite.loaded == true){
           this.sprite.position = this.position;
        }
    }

    /**
     * Callback called to render the object by the renderer
     */
    draw(context: Context): void {
        if(this.sprite.loaded == true){
            context.drawSprite(this.sprite)
        }
    }

    /**
     * Register gameobject in the renderingLayer
     * @param layer
     */
    registerToLayer(layer: RenderLayer): void {
        this.uuid = layer.addObject(this);
        this.renderLayer = layer;
    }


}