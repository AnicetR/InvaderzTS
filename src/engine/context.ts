import {boundaries} from './types/contextTypes';
import { Sprite } from './baseObjects/sprite';

/**
 * engine/context.ts
 * Manages the context
 */
export class Context {
    
    public boundaries: boundaries;
    private readonly context: CanvasRenderingContext2D;

    /**
     * constructor
     * @param canvas The canvas element to render to
     */
    constructor(private readonly canvas: HTMLCanvasElement){
        this.context = this.canvas.getContext('2d');

        this.boundaries = {
            maxX : this.canvas.height,
            maxY : this.canvas.width
        };
    }

    /**
     * returns the current canvas context
     */
    public get() : CanvasRenderingContext2D {
        return this.context;
    }

    /**
     * save the state of the canvas context in a stack
     */
    public save() : void {
        this.context.save();
    }

    /**
     * Clears the canvas content
     */
    public clear() : void{
        this.context.clearRect(0, 0, this.context.canvas.width, this.context.canvas.height);
    }

    /**
     * Draw a sprite into context
     * @param sprite Sprite to draw
     */
    public drawSprite(sprite: Sprite) : void {
        this.context.drawImage(sprite, sprite.x, sprite.y, sprite.width, sprite.height);
    }
}

