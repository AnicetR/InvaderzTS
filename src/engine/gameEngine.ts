import { Context } from "./context";

export class gameEngine{
    private readonly context : Context;

    constructor(private readonly canvas: HTMLCanvasElement){
        this.context = new Context(this.canvas);
    }
}