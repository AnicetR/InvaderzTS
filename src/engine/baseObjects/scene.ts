import { sceneInterface} from "../types/scenesTypes";
import { RenderLayerInterface } from "../types/renderLayerInterface";
import { gameEngine } from "../gameEngine";
import { intervalInterface } from "../types/intervalTypes";


export class scene implements sceneInterface{
    _layersCollection: Map<string, RenderLayerInterface> = new Map();
    _timeoutsCollection: Map<string, intervalInterface> = new Map();

    constructor(readonly _gameEngine: gameEngine){}

    load(): void {
        for(const [layerName, layer] of this._layersCollection){
            this._gameEngine.renderer.addLayer(layer)
        }
    }

    unload(): void {
        for(const [layerName, layer] of this._layersCollection){
            console.log(layer.name)
            this._gameEngine.renderer.removeLayer(layer)
        }
        for(const [timeoutName, timeout] of this._timeoutsCollection){
            timeout.stop();
        }
    }
}