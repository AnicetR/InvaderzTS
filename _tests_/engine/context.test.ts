import { Context } from "../../src/engine/context";
import { Sprite } from "../../src/engine/baseObjects/sprite";

const canvas: HTMLCanvasElement = document.createElement('canvas');
canvas.height = 300;
canvas.width = 150;
const context = new Context(canvas);

test('context.construct', () => {
    expect(context).toBeInstanceOf(Context)
})


test('context.boundaries', () => {
    expect(context.boundaries)
    .toStrictEqual(
        {"maxX": 300, "maxY": 150}
    )
})

test('context.getCtx', () => {
    expect(context.get()).toBe(canvas.getContext("2d"))
})

test('context.managerFunction(.save, .restore, .clear)', () => {
    //@ts-ignore
    const events = context.get().__getEvents();
    context.save();
    context.clear();
    context.restore();
    expect(context.get()).toMatchSnapshot();
}) 

test('context.drawSprite', () => {
    //@ts-ignore
    const calls  = context.get().__getDrawCalls();
    const imageBase64 = "data:image/jpeg;base64,/9j/4AAQSkZJRgABAQEA3ADcAAD/2wBDAAIBAQIBAQICAgICAgICAwUDAwMDAwYEBAMFBwYHBwcGBwcICQsJCAgKCAcHCg0KCgsMDAwMBwkODw0MDgsMDAz/2wBDAQICAgMDAwYDAwYMCAcIDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAz/wgARCAABAAEDAREAAhEBAxEB/8QAFAABAAAAAAAAAAAAAAAAAAAACP/EABQBAQAAAAAAAAAAAAAAAAAAAAD/2gAMAwEAAhADEAAAAX8f/8QAFBABAAAAAAAAAAAAAAAAAAAAAP/aAAgBAQABBQJ//8QAFBEBAAAAAAAAAAAAAAAAAAAAAP/aAAgBAwEBPwF//8QAFBEBAAAAAAAAAAAAAAAAAAAAAP/aAAgBAgEBPwF//8QAFBABAAAAAAAAAAAAAAAAAAAAAP/aAAgBAQAGPwJ//8QAFBABAAAAAAAAAAAAAAAAAAAAAP/aAAgBAQABPyF//9oADAMBAAIAAwAAABAf/8QAFBEBAAAAAAAAAAAAAAAAAAAAAP/aAAgBAwEBPxB//8QAFBEBAAAAAAAAAAAAAAAAAAAAAP/aAAgBAgEBPxB//8QAFBABAAAAAAAAAAAAAAAAAAAAAP/aAAgBAQABPxB//9k=";
    const sprite = new Sprite(imageBase64);
    context.drawSprite(sprite);
    expect(context.get()).toMatchSnapshot();
})



