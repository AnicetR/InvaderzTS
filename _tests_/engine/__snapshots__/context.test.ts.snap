// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`context.drawSprite 1`] = `
CanvasRenderingContext2D {
  "_canvas": <canvas
    height="300"
    width="150"
  />,
  "_directionStack": Array [
    "inherit",
  ],
  "_drawCalls": Array [
    Object {
      "props": Object {
        "height": 300,
        "width": 150,
        "x": 0,
        "y": 0,
      },
      "transform": Array [
        1,
        0,
        0,
        1,
        0,
        0,
      ],
      "type": "clearRect",
    },
    Object {
      "props": Object {
        "dHeight": 0,
        "dWidth": 0,
        "dx": 0,
        "dy": 0,
        "img": <img
          src="data:image/jpeg;base64,/9j/4AAQSkZJRgABAQEA3ADcAAD/2wBDAAIBAQIBAQICAgICAgICAwUDAwMDAwYEBAMFBwYHBwcGBwcICQsJCAgKCAcHCg0KCgsMDAwMBwkODw0MDgsMDAz/2wBDAQICAgMDAwYDAwYMCAcIDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAz/wgARCAABAAEDAREAAhEBAxEB/8QAFAABAAAAAAAAAAAAAAAAAAAACP/EABQBAQAAAAAAAAAAAAAAAAAAAAD/2gAMAwEAAhADEAAAAX8f/8QAFBABAAAAAAAAAAAAAAAAAAAAAP/aAAgBAQABBQJ//8QAFBEBAAAAAAAAAAAAAAAAAAAAAP/aAAgBAwEBPwF//8QAFBEBAAAAAAAAAAAAAAAAAAAAAP/aAAgBAgEBPwF//8QAFBABAAAAAAAAAAAAAAAAAAAAAP/aAAgBAQAGPwJ//8QAFBABAAAAAAAAAAAAAAAAAAAAAP/aAAgBAQABPyF//9oADAMBAAIAAwAAABAf/8QAFBEBAAAAAAAAAAAAAAAAAAAAAP/aAAgBAwEBPxB//8QAFBEBAAAAAAAAAAAAAAAAAAAAAP/aAAgBAgEBPxB//8QAFBABAAAAAAAAAAAAAAAAAAAAAP/aAAgBAQABPxB//9k="
        />,
        "sHeight": 0,
        "sWidth": 0,
        "sx": 0,
        "sy": 0,
      },
      "transform": Array [
        1,
        0,
        0,
        1,
        0,
        0,
      ],
      "type": "drawImage",
    },
  ],
  "_events": Array [
    Object {
      "props": Object {},
      "transform": Array [
        1,
        0,
        0,
        1,
        0,
        0,
      ],
      "type": "save",
    },
    Object {
      "props": Object {
        "height": 300,
        "width": 150,
        "x": 0,
        "y": 0,
      },
      "transform": Array [
        1,
        0,
        0,
        1,
        0,
        0,
      ],
      "type": "clearRect",
    },
    Object {
      "props": Object {},
      "transform": Array [
        1,
        0,
        0,
        1,
        0,
        0,
      ],
      "type": "restore",
    },
    Object {
      "props": Object {
        "dHeight": 0,
        "dWidth": 0,
        "dx": 0,
        "dy": 0,
        "img": <img
          src="data:image/jpeg;base64,/9j/4AAQSkZJRgABAQEA3ADcAAD/2wBDAAIBAQIBAQICAgICAgICAwUDAwMDAwYEBAMFBwYHBwcGBwcICQsJCAgKCAcHCg0KCgsMDAwMBwkODw0MDgsMDAz/2wBDAQICAgMDAwYDAwYMCAcIDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAz/wgARCAABAAEDAREAAhEBAxEB/8QAFAABAAAAAAAAAAAAAAAAAAAACP/EABQBAQAAAAAAAAAAAAAAAAAAAAD/2gAMAwEAAhADEAAAAX8f/8QAFBABAAAAAAAAAAAAAAAAAAAAAP/aAAgBAQABBQJ//8QAFBEBAAAAAAAAAAAAAAAAAAAAAP/aAAgBAwEBPwF//8QAFBEBAAAAAAAAAAAAAAAAAAAAAP/aAAgBAgEBPwF//8QAFBABAAAAAAAAAAAAAAAAAAAAAP/aAAgBAQAGPwJ//8QAFBABAAAAAAAAAAAAAAAAAAAAAP/aAAgBAQABPyF//9oADAMBAAIAAwAAABAf/8QAFBEBAAAAAAAAAAAAAAAAAAAAAP/aAAgBAwEBPxB//8QAFBEBAAAAAAAAAAAAAAAAAAAAAP/aAAgBAgEBPxB//8QAFBABAAAAAAAAAAAAAAAAAAAAAP/aAAgBAQABPxB//9k="
        />,
        "sHeight": 0,
        "sWidth": 0,
        "sx": 0,
        "sy": 0,
      },
      "transform": Array [
        1,
        0,
        0,
        1,
        0,
        0,
      ],
      "type": "drawImage",
    },
  ],
  "_fillStyleStack": Array [
    "#000",
  ],
  "_filterStack": Array [
    "none",
  ],
  "_fontStack": Array [
    "10px sans-serif",
  ],
  "_globalAlphaStack": Array [
    1,
  ],
  "_globalCompositeOperationStack": Array [
    "source-over",
  ],
  "_imageSmoothingEnabledStack": Array [
    true,
  ],
  "_imageSmoothingQualityStack": Array [
    "low",
  ],
  "_lineCapStack": Array [
    "butt",
  ],
  "_lineDashOffsetStack": Array [
    0,
  ],
  "_lineDashStack": Array [
    Array [],
  ],
  "_lineJoinStack": Array [
    "miter",
  ],
  "_lineWidthStack": Array [
    1,
  ],
  "_miterLimitStack": Array [
    10,
  ],
  "_path": Array [
    Object {
      "props": Object {},
      "transform": Array [
        1,
        0,
        0,
        1,
        0,
        0,
      ],
      "type": "beginPath",
    },
  ],
  "_shadowBlurStack": Array [
    0,
  ],
  "_shadowColorStack": Array [
    "rgba(0, 0, 0, 0)",
  ],
  "_shadowOffsetXStack": Array [
    0,
  ],
  "_shadowOffsetYStack": Array [
    0,
  ],
  "_stackIndex": 0,
  "_strokeStyleStack": Array [
    "#000",
  ],
  "_textAlignStack": Array [
    "start",
  ],
  "_textBaselineStack": Array [
    "alphabetic",
  ],
  "_transformStack": Array [
    Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
  ],
  "addHitRegion": [MockFunction],
  "arc": [MockFunction],
  "arcTo": [MockFunction],
  "beginPath": [MockFunction],
  "bezierCurveTo": [MockFunction],
  "clearHitRegions": [MockFunction],
  "clearRect": [MockFunction] {
    "calls": Array [
      Array [
        0,
        0,
        150,
        300,
      ],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  "clip": [MockFunction],
  "closePath": [MockFunction],
  "createImageData": [MockFunction],
  "createLinearGradient": [MockFunction],
  "createPattern": [MockFunction],
  "createRadialGradient": [MockFunction],
  "drawFocusIfNeeded": [MockFunction],
  "drawImage": [MockFunction] {
    "calls": Array [
      Array [
        <img
          src="data:image/jpeg;base64,/9j/4AAQSkZJRgABAQEA3ADcAAD/2wBDAAIBAQIBAQICAgICAgICAwUDAwMDAwYEBAMFBwYHBwcGBwcICQsJCAgKCAcHCg0KCgsMDAwMBwkODw0MDgsMDAz/2wBDAQICAgMDAwYDAwYMCAcIDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAz/wgARCAABAAEDAREAAhEBAxEB/8QAFAABAAAAAAAAAAAAAAAAAAAACP/EABQBAQAAAAAAAAAAAAAAAAAAAAD/2gAMAwEAAhADEAAAAX8f/8QAFBABAAAAAAAAAAAAAAAAAAAAAP/aAAgBAQABBQJ//8QAFBEBAAAAAAAAAAAAAAAAAAAAAP/aAAgBAwEBPwF//8QAFBEBAAAAAAAAAAAAAAAAAAAAAP/aAAgBAgEBPwF//8QAFBABAAAAAAAAAAAAAAAAAAAAAP/aAAgBAQAGPwJ//8QAFBABAAAAAAAAAAAAAAAAAAAAAP/aAAgBAQABPyF//9oADAMBAAIAAwAAABAf/8QAFBEBAAAAAAAAAAAAAAAAAAAAAP/aAAgBAwEBPxB//8QAFBEBAAAAAAAAAAAAAAAAAAAAAP/aAAgBAgEBPxB//8QAFBABAAAAAAAAAAAAAAAAAAAAAP/aAAgBAQABPxB//9k="
        />,
        0,
        0,
        0,
        0,
      ],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  "ellipse": [MockFunction],
  "fill": [MockFunction],
  "fillRect": [MockFunction],
  "fillText": [MockFunction],
  "getImageData": [MockFunction],
  "getLineDash": [MockFunction],
  "getTransform": [MockFunction],
  "isPointInPath": [MockFunction],
  "isPointInStroke": [MockFunction],
  "lineTo": [MockFunction],
  "measureText": [MockFunction],
  "moveTo": [MockFunction],
  "putImageData": [MockFunction],
  "quadraticCurveTo": [MockFunction],
  "rect": [MockFunction],
  "removeHitRegion": [MockFunction],
  "resetTransform": [MockFunction],
  "restore": [MockFunction] {
    "calls": Array [
      Array [],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  "rotate": [MockFunction],
  "save": [MockFunction] {
    "calls": Array [
      Array [],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  "scale": [MockFunction],
  "scrollPathIntoView": [MockFunction],
  "setLineDash": [MockFunction],
  "setTransform": [MockFunction],
  "stroke": [MockFunction],
  "strokeRect": [MockFunction],
  "strokeText": [MockFunction],
  "transform": [MockFunction],
  "translate": [MockFunction],
}
`;

exports[`context.managerFunction(.save, .restore, .clear) 1`] = `
CanvasRenderingContext2D {
  "_canvas": <canvas
    height="300"
    width="150"
  />,
  "_directionStack": Array [
    "inherit",
  ],
  "_drawCalls": Array [
    Object {
      "props": Object {
        "height": 300,
        "width": 150,
        "x": 0,
        "y": 0,
      },
      "transform": Array [
        1,
        0,
        0,
        1,
        0,
        0,
      ],
      "type": "clearRect",
    },
  ],
  "_events": Array [
    Object {
      "props": Object {},
      "transform": Array [
        1,
        0,
        0,
        1,
        0,
        0,
      ],
      "type": "save",
    },
    Object {
      "props": Object {
        "height": 300,
        "width": 150,
        "x": 0,
        "y": 0,
      },
      "transform": Array [
        1,
        0,
        0,
        1,
        0,
        0,
      ],
      "type": "clearRect",
    },
    Object {
      "props": Object {},
      "transform": Array [
        1,
        0,
        0,
        1,
        0,
        0,
      ],
      "type": "restore",
    },
  ],
  "_fillStyleStack": Array [
    "#000",
  ],
  "_filterStack": Array [
    "none",
  ],
  "_fontStack": Array [
    "10px sans-serif",
  ],
  "_globalAlphaStack": Array [
    1,
  ],
  "_globalCompositeOperationStack": Array [
    "source-over",
  ],
  "_imageSmoothingEnabledStack": Array [
    true,
  ],
  "_imageSmoothingQualityStack": Array [
    "low",
  ],
  "_lineCapStack": Array [
    "butt",
  ],
  "_lineDashOffsetStack": Array [
    0,
  ],
  "_lineDashStack": Array [
    Array [],
  ],
  "_lineJoinStack": Array [
    "miter",
  ],
  "_lineWidthStack": Array [
    1,
  ],
  "_miterLimitStack": Array [
    10,
  ],
  "_path": Array [
    Object {
      "props": Object {},
      "transform": Array [
        1,
        0,
        0,
        1,
        0,
        0,
      ],
      "type": "beginPath",
    },
  ],
  "_shadowBlurStack": Array [
    0,
  ],
  "_shadowColorStack": Array [
    "rgba(0, 0, 0, 0)",
  ],
  "_shadowOffsetXStack": Array [
    0,
  ],
  "_shadowOffsetYStack": Array [
    0,
  ],
  "_stackIndex": 0,
  "_strokeStyleStack": Array [
    "#000",
  ],
  "_textAlignStack": Array [
    "start",
  ],
  "_textBaselineStack": Array [
    "alphabetic",
  ],
  "_transformStack": Array [
    Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
  ],
  "addHitRegion": [MockFunction],
  "arc": [MockFunction],
  "arcTo": [MockFunction],
  "beginPath": [MockFunction],
  "bezierCurveTo": [MockFunction],
  "clearHitRegions": [MockFunction],
  "clearRect": [MockFunction] {
    "calls": Array [
      Array [
        0,
        0,
        150,
        300,
      ],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  "clip": [MockFunction],
  "closePath": [MockFunction],
  "createImageData": [MockFunction],
  "createLinearGradient": [MockFunction],
  "createPattern": [MockFunction],
  "createRadialGradient": [MockFunction],
  "drawFocusIfNeeded": [MockFunction],
  "drawImage": [MockFunction],
  "ellipse": [MockFunction],
  "fill": [MockFunction],
  "fillRect": [MockFunction],
  "fillText": [MockFunction],
  "getImageData": [MockFunction],
  "getLineDash": [MockFunction],
  "getTransform": [MockFunction],
  "isPointInPath": [MockFunction],
  "isPointInStroke": [MockFunction],
  "lineTo": [MockFunction],
  "measureText": [MockFunction],
  "moveTo": [MockFunction],
  "putImageData": [MockFunction],
  "quadraticCurveTo": [MockFunction],
  "rect": [MockFunction],
  "removeHitRegion": [MockFunction],
  "resetTransform": [MockFunction],
  "restore": [MockFunction] {
    "calls": Array [
      Array [],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  "rotate": [MockFunction],
  "save": [MockFunction] {
    "calls": Array [
      Array [],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  "scale": [MockFunction],
  "scrollPathIntoView": [MockFunction],
  "setLineDash": [MockFunction],
  "setTransform": [MockFunction],
  "stroke": [MockFunction],
  "strokeRect": [MockFunction],
  "strokeText": [MockFunction],
  "transform": [MockFunction],
  "translate": [MockFunction],
}
`;
